File: frontend/src/types/workspace_management.ts
/**
 * Workspace Management Types
 *
 * TypeScript types for multi-tenant workspace management, member management,
 * invitations, and workspace settings.
 */

// Enums

export enum WorkspaceStatus {
  ACTIVE = 'active',
  SUSPENDED = 'suspended',
  TRIAL = 'trial',
  EXPIRED = 'expired',
  DELETED = 'deleted',
}

export enum WorkspacePlan {
  FREE = 'free',
  STARTER = 'starter',
  PRO = 'pro',
  BUSINESS = 'business',
  ENTERPRISE = 'enterprise',
}

export enum MemberRole {
  OWNER = 'owner',
  ADMIN = 'admin',
  EDITOR = 'editor',
  VIEWER = 'viewer',
  GUEST = 'guest',
}

export enum MemberStatus {
  ACTIVE = 'active',
  INVITED = 'invited',
  SUSPENDED = 'suspended',
  DEACTIVATED = 'deactivated',
}

export enum InvitationStatus {
  PENDING = 'pending',
  ACCEPTED = 'accepted',
  DECLINED = 'declined',
  EXPIRED = 'expired',
  CANCELLED = 'cancelled',
}

export enum ResourceType {
  DASHBOARD = 'dashboard',
  CHART = 'chart',
  CONNECTION = 'connection',
  DATASET = 'dataset',
  TRANSFORM = 'transform',
  SEMANTIC_MODEL = 'semantic_model',
  QUERY = 'query',
  KPI = 'kpi',
  REPORT = 'report',
}

// Workspace Interfaces

export interface WorkspaceCreate {
  name: string;
  slug: string;
  description?: string;
  logo_url?: string;
  plan?: WorkspacePlan;
  settings?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface Workspace {
  id: string;
  name: string;
  slug: string;
  description?: string;
  logo_url?: string;
  status: WorkspaceStatus;
  plan: WorkspacePlan;
  owner_id: string;
  member_count: number;
  dashboard_count: number;
  connection_count: number;
  storage_used_mb: number;
  settings: Record<string, unknown>;
  metadata: Record<string, unknown>;
  trial_ends_at?: string;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceUpdate {
  name?: string;
  description?: string;
  logo_url?: string;
  settings?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface WorkspaceSettings {
  workspace_id: string;
  allow_public_dashboards: boolean;
  allow_public_links: boolean;
  require_2fa: boolean;
  default_member_role: MemberRole;
  allowed_email_domains: string[];
  ip_whitelist: string[];
  session_timeout_minutes: number;
  password_policy: Record<string, unknown>;
  branding: Record<string, unknown>;
  notification_settings: Record<string, unknown>;
  data_retention_days: number;
  max_connections: number;
  max_dashboards: number;
  max_members: number;
  features_enabled: string[];
}

export interface WorkspaceSettingsUpdate {
  allow_public_dashboards?: boolean;
  allow_public_links?: boolean;
  require_2fa?: boolean;
  default_member_role?: MemberRole;
  allowed_email_domains?: string[];
  ip_whitelist?: string[];
  session_timeout_minutes?: number;
  password_policy?: Record<string, unknown>;
  branding?: Record<string, unknown>;
  notification_settings?: Record<string, unknown>;
  data_retention_days?: number;
}

// Member Interfaces

export interface WorkspaceMemberCreate {
  user_id: string;
  role?: MemberRole;
  permissions?: Record<string, boolean>;
}

export interface WorkspaceMember {
  id: string;
  workspace_id: string;
  user_id: string;
  email: string;
  name: string;
  avatar_url?: string;
  role: MemberRole;
  status: MemberStatus;
  permissions: Record<string, boolean>;
  last_active_at?: string;
  joined_at: string;
  invited_by?: string;
}

export interface WorkspaceMemberUpdate {
  role?: MemberRole;
  permissions?: Record<string, boolean>;
  status?: MemberStatus;
}

// Invitation Interfaces

export interface InvitationCreate {
  email: string;
  role?: MemberRole;
  message?: string;
  expires_in_days?: number;
}

export interface Invitation {
  id: string;
  workspace_id: string;
  workspace_name: string;
  email: string;
  role: MemberRole;
  status: InvitationStatus;
  message?: string;
  token: string;
  invited_by_id: string;
  invited_by_name: string;
  expires_at: string;
  accepted_at?: string;
  declined_at?: string;
  created_at: string;
}

export interface BulkInvitationCreate {
  emails: string[];
  role?: MemberRole;
  message?: string;
  expires_in_days?: number;
}

export interface InvitationResponse {
  accept: boolean;
  user_id?: string;
}

// Quota Interfaces

export interface WorkspaceQuota {
  workspace_id: string;
  plan: WorkspacePlan;
  max_members: number;
  current_members: number;
  max_guests: number;
  current_guests: number;
  max_connections: number;
  current_connections: number;
  max_dashboards: number;
  current_dashboards: number;
  max_charts: number;
  current_charts: number;
  max_datasets: number;
  current_datasets: number;
  max_storage_mb: number;
  current_storage_mb: number;
  max_api_calls_per_day: number;
  current_api_calls_today: number;
  max_query_execution_minutes: number;
  current_query_minutes_today: number;
  max_scheduled_reports: number;
  current_scheduled_reports: number;
  max_alerts: number;
  current_alerts: number;
}

export interface QuotaUsage {
  workspace_id: string;
  resource_type: string;
  limit: number;
  used: number;
  remaining: number;
  percent_used: number;
}

// Access Control Interfaces

export interface ResourcePermission {
  resource_type: ResourceType;
  resource_id: string;
  member_id: string;
  can_view: boolean;
  can_edit: boolean;
  can_delete: boolean;
  can_share: boolean;
  can_export: boolean;
}

export interface RolePermissions {
  role: MemberRole;
  can_manage_workspace: boolean;
  can_manage_members: boolean;
  can_manage_billing: boolean;
  can_manage_settings: boolean;
  can_create_connections: boolean;
  can_create_dashboards: boolean;
  can_create_charts: boolean;
  can_create_datasets: boolean;
  can_execute_queries: boolean;
  can_export_data: boolean;
  can_share_resources: boolean;
  can_create_public_links: boolean;
  can_use_ai_features: boolean;
  can_create_alerts: boolean;
  can_schedule_reports: boolean;
  can_use_api: boolean;
}

// Team Interfaces

export interface Team {
  id: string;
  workspace_id: string;
  name: string;
  description?: string;
  color?: string;
  member_ids: string[];
  permissions: Record<string, boolean>;
  created_at: string;
}

export interface TeamCreate {
  name: string;
  description?: string;
  color?: string;
  member_ids?: string[];
}

export interface TeamUpdate {
  name?: string;
  description?: string;
  color?: string;
  member_ids?: string[];
  permissions?: Record<string, boolean>;
}

// Activity Interfaces

export interface WorkspaceActivity {
  id: string;
  workspace_id: string;
  user_id: string;
  user_name: string;
  action: string;
  resource_type?: ResourceType;
  resource_id?: string;
  resource_name?: string;
  details: Record<string, unknown>;
  ip_address?: string;
  user_agent?: string;
  created_at: string;
}

// Response Interfaces

export interface WorkspaceListResponse {
  workspaces: Workspace[];
  total: number;
}

export interface WorkspaceMemberListResponse {
  members: WorkspaceMember[];
  total: number;
}

export interface InvitationListResponse {
  invitations: Invitation[];
  total: number;
}

export interface TeamListResponse {
  teams: Team[];
  total: number;
}

export interface WorkspaceActivityListResponse {
  activities: WorkspaceActivity[];
  total: number;
}

export interface QuotaUsageListResponse {
  quotas: QuotaUsage[];
}

// Summary Interfaces

export interface WorkspaceSummary {
  workspace_id: string;
  name: string;
  plan: WorkspacePlan;
  status: WorkspaceStatus;
  member_count: number;
  dashboard_count: number;
  connection_count: number;
  chart_count: number;
  storage_used_mb: number;
  storage_limit_mb: number;
  api_calls_today: number;
  active_users_today: number;
  recent_activities: WorkspaceActivity[];
}

export interface UserWorkspaces {
  owned: Workspace[];
  member_of: Workspace[];
  pending_invitations: Invitation[];
}

// Permission Check Interface

export interface PermissionCheck {
  workspace_id: string;
  member_id: string;
  permission: string;
  allowed: boolean;
  reason?: string;
}

// Constants

export const PLAN_LIMITS: Record<WorkspacePlan, Record<string, number>> = {
  [WorkspacePlan.FREE]: {
    max_members: 3,
    max_guests: 0,
    max_connections: 2,
    max_dashboards: 5,
    max_charts: 20,
    max_datasets: 5,
    max_storage_mb: 100,
    max_api_calls_per_day: 100,
    max_query_execution_minutes: 10,
    max_scheduled_reports: 0,
    max_alerts: 0,
  },
  [WorkspacePlan.STARTER]: {
    max_members: 5,
    max_guests: 5,
    max_connections: 5,
    max_dashboards: 20,
    max_charts: 100,
    max_datasets: 20,
    max_storage_mb: 500,
    max_api_calls_per_day: 1000,
    max_query_execution_minutes: 60,
    max_scheduled_reports: 5,
    max_alerts: 10,
  },
  [WorkspacePlan.PRO]: {
    max_members: 25,
    max_guests: 25,
    max_connections: 20,
    max_dashboards: 100,
    max_charts: 500,
    max_datasets: 100,
    max_storage_mb: 5000,
    max_api_calls_per_day: 10000,
    max_query_execution_minutes: 300,
    max_scheduled_reports: 25,
    max_alerts: 50,
  },
  [WorkspacePlan.BUSINESS]: {
    max_members: 100,
    max_guests: 100,
    max_connections: 50,
    max_dashboards: 500,
    max_charts: 2500,
    max_datasets: 500,
    max_storage_mb: 25000,
    max_api_calls_per_day: 100000,
    max_query_execution_minutes: 1000,
    max_scheduled_reports: 100,
    max_alerts: 200,
  },
  [WorkspacePlan.ENTERPRISE]: {
    max_members: -1,
    max_guests: -1,
    max_connections: -1,
    max_dashboards: -1,
    max_charts: -1,
    max_datasets: -1,
    max_storage_mb: -1,
    max_api_calls_per_day: -1,
    max_query_execution_minutes: -1,
    max_scheduled_reports: -1,
    max_alerts: -1,
  },
};

export const ROLE_HIERARCHY: MemberRole[] = [
  MemberRole.OWNER,
  MemberRole.ADMIN,
  MemberRole.EDITOR,
  MemberRole.VIEWER,
  MemberRole.GUEST,
];

export const ROLE_LABELS: Record<MemberRole, string> = {
  [MemberRole.OWNER]: 'Owner',
  [MemberRole.ADMIN]: 'Admin',
  [MemberRole.EDITOR]: 'Editor',
  [MemberRole.VIEWER]: 'Viewer',
  [MemberRole.GUEST]: 'Guest',
};

export const PLAN_LABELS: Record<WorkspacePlan, string> = {
  [WorkspacePlan.FREE]: 'Free',
  [WorkspacePlan.STARTER]: 'Starter',
  [WorkspacePlan.PRO]: 'Pro',
  [WorkspacePlan.BUSINESS]: 'Business',
  [WorkspacePlan.ENTERPRISE]: 'Enterprise',
};

export const STATUS_LABELS: Record<WorkspaceStatus, string> = {
  [WorkspaceStatus.ACTIVE]: 'Active',
  [WorkspaceStatus.SUSPENDED]: 'Suspended',
  [WorkspaceStatus.TRIAL]: 'Trial',
  [WorkspaceStatus.EXPIRED]: 'Expired',
  [WorkspaceStatus.DELETED]: 'Deleted',
};

export const INVITATION_STATUS_LABELS: Record<InvitationStatus, string> = {
  [InvitationStatus.PENDING]: 'Pending',
  [InvitationStatus.ACCEPTED]: 'Accepted',
  [InvitationStatus.DECLINED]: 'Declined',
  [InvitationStatus.EXPIRED]: 'Expired',
  [InvitationStatus.CANCELLED]: 'Cancelled',
};

export const MEMBER_STATUS_LABELS: Record<MemberStatus, string> = {
  [MemberStatus.ACTIVE]: 'Active',
  [MemberStatus.INVITED]: 'Invited',
  [MemberStatus.SUSPENDED]: 'Suspended',
  [MemberStatus.DEACTIVATED]: 'Deactivated',
};

export const RESOURCE_TYPE_LABELS: Record<ResourceType, string> = {
  [ResourceType.DASHBOARD]: 'Dashboard',
  [ResourceType.CHART]: 'Chart',
  [ResourceType.CONNECTION]: 'Connection',
  [ResourceType.DATASET]: 'Dataset',
  [ResourceType.TRANSFORM]: 'Transform',
  [ResourceType.SEMANTIC_MODEL]: 'Semantic Model',
  [ResourceType.QUERY]: 'Query',
  [ResourceType.KPI]: 'KPI',
  [ResourceType.REPORT]: 'Report',
};

// Helper Functions

export function getPlanLimits(plan: WorkspacePlan): Record<string, number> {
  return PLAN_LIMITS[plan] || PLAN_LIMITS[WorkspacePlan.FREE];
}

export function isUnlimited(limit: number): boolean {
  return limit === -1;
}

export function formatQuotaValue(value: number, limit: number): string {
  if (isUnlimited(limit)) {
    return `${value} / Unlimited`;
  }
  return `${value} / ${limit}`;
}

export function calculateQuotaPercentage(used: number, limit: number): number {
  if (isUnlimited(limit)) return 0;
  if (limit === 0) return 100;
  return Math.min(100, Math.round((used / limit) * 100));
}

export function getQuotaStatus(
  used: number,
  limit: number
): 'ok' | 'warning' | 'critical' {
  if (isUnlimited(limit)) return 'ok';
  const percentage = calculateQuotaPercentage(used, limit);
  if (percentage >= 100) return 'critical';
  if (percentage >= 80) return 'warning';
  return 'ok';
}

export function canRoleManage(managerRole: MemberRole, targetRole: MemberRole): boolean {
  const managerIndex = ROLE_HIERARCHY.indexOf(managerRole);
  const targetIndex = ROLE_HIERARCHY.indexOf(targetRole);
  return managerIndex < targetIndex;
}

export function isRoleAtLeast(role: MemberRole, minimumRole: MemberRole): boolean {
  const roleIndex = ROLE_HIERARCHY.indexOf(role);
  const minimumIndex = ROLE_HIERARCHY.indexOf(minimumRole);
  return roleIndex <= minimumIndex;
}

export function formatStorageSize(megabytes: number): string {
  if (megabytes < 1024) {
    return `${megabytes} MB`;
  }
  return `${(megabytes / 1024).toFixed(1)} GB`;
}

export function isInvitationExpired(invitation: Invitation): boolean {
  return new Date(invitation.expires_at) < new Date();
}

export function isPendingInvitation(invitation: Invitation): boolean {
  return invitation.status === InvitationStatus.PENDING && !isInvitationExpired(invitation);
}

export function getActivityIcon(action: string): string {
  const icons: Record<string, string> = {
    created: '➕',
    updated: '✏️',
    deleted: '🗑️',
    invited: '✉️',
    joined: '👋',
    left: '👋',
    shared: '🔗',
    exported: '📤',
    imported: '📥',
  };
  return icons[action] || '📝';
}

export function formatActivityMessage(activity: WorkspaceActivity): string {
  const { user_name, action, resource_type, resource_name } = activity;

  if (resource_type && resource_name) {
    return `${user_name} ${action} ${RESOURCE_TYPE_LABELS[resource_type].toLowerCase()} "${resource_name}"`;
  }

  return `${user_name} ${action}`;
}

File: backend/app/schemas/workspace_management.py
"""
Workspace Management Schemas

Pydantic schemas for multi-tenant workspace management, member management,
invitations, and workspace settings.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field, EmailStr
from enum import Enum
from datetime import datetime


# Enums

class WorkspaceStatus(str, Enum):
    """Workspace status"""
    ACTIVE = "active"
    SUSPENDED = "suspended"
    TRIAL = "trial"
    EXPIRED = "expired"
    DELETED = "deleted"


class WorkspacePlan(str, Enum):
    """Workspace subscription plan"""
    FREE = "free"
    STARTER = "starter"
    PRO = "pro"
    BUSINESS = "business"
    ENTERPRISE = "enterprise"


class MemberRole(str, Enum):
    """Member role within workspace"""
    OWNER = "owner"
    ADMIN = "admin"
    EDITOR = "editor"
    VIEWER = "viewer"
    GUEST = "guest"


class MemberStatus(str, Enum):
    """Member status"""
    ACTIVE = "active"
    INVITED = "invited"
    SUSPENDED = "suspended"
    DEACTIVATED = "deactivated"


class InvitationStatus(str, Enum):
    """Invitation status"""
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    EXPIRED = "expired"
    CANCELLED = "cancelled"


class ResourceType(str, Enum):
    """Types of workspace resources"""
    DASHBOARD = "dashboard"
    CHART = "chart"
    CONNECTION = "connection"
    DATASET = "dataset"
    TRANSFORM = "transform"
    SEMANTIC_MODEL = "semantic_model"
    QUERY = "query"
    KPI = "kpi"
    REPORT = "report"


# Workspace Models

class WorkspaceCreate(BaseModel):
    """Create a new workspace"""
    name: str
    slug: str
    description: Optional[str] = None
    logo_url: Optional[str] = None
    plan: WorkspacePlan = WorkspacePlan.FREE
    settings: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class Workspace(BaseModel):
    """Workspace model"""
    id: str
    name: str
    slug: str
    description: Optional[str] = None
    logo_url: Optional[str] = None
    status: WorkspaceStatus
    plan: WorkspacePlan
    owner_id: str
    member_count: int = 0
    dashboard_count: int = 0
    connection_count: int = 0
    storage_used_mb: float = 0
    settings: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)
    trial_ends_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class WorkspaceUpdate(BaseModel):
    """Update workspace fields"""
    name: Optional[str] = None
    description: Optional[str] = None
    logo_url: Optional[str] = None
    settings: Optional[dict[str, Any]] = None
    metadata: Optional[dict[str, Any]] = None


class WorkspaceSettings(BaseModel):
    """Workspace settings"""
    workspace_id: str
    allow_public_dashboards: bool = True
    allow_public_links: bool = True
    require_2fa: bool = False
    default_member_role: MemberRole = MemberRole.VIEWER
    allowed_email_domains: list[str] = Field(default_factory=list)
    ip_whitelist: list[str] = Field(default_factory=list)
    session_timeout_minutes: int = 480  # 8 hours
    password_policy: dict[str, Any] = Field(default_factory=dict)
    branding: dict[str, Any] = Field(default_factory=dict)
    notification_settings: dict[str, Any] = Field(default_factory=dict)
    data_retention_days: int = 365
    max_connections: int = 10
    max_dashboards: int = 50
    max_members: int = 10
    features_enabled: list[str] = Field(default_factory=list)


class WorkspaceSettingsUpdate(BaseModel):
    """Update workspace settings"""
    allow_public_dashboards: Optional[bool] = None
    allow_public_links: Optional[bool] = None
    require_2fa: Optional[bool] = None
    default_member_role: Optional[MemberRole] = None
    allowed_email_domains: Optional[list[str]] = None
    ip_whitelist: Optional[list[str]] = None
    session_timeout_minutes: Optional[int] = None
    password_policy: Optional[dict[str, Any]] = None
    branding: Optional[dict[str, Any]] = None
    notification_settings: Optional[dict[str, Any]] = None
    data_retention_days: Optional[int] = None


# Member Models

class WorkspaceMemberCreate(BaseModel):
    """Add member to workspace"""
    user_id: str
    role: MemberRole = MemberRole.VIEWER
    permissions: dict[str, bool] = Field(default_factory=dict)


class WorkspaceMember(BaseModel):
    """Workspace member"""
    id: str
    workspace_id: str
    user_id: str
    email: str
    name: str
    avatar_url: Optional[str] = None
    role: MemberRole
    status: MemberStatus
    permissions: dict[str, bool] = Field(default_factory=dict)
    last_active_at: Optional[datetime] = None
    joined_at: datetime = Field(default_factory=datetime.utcnow)
    invited_by: Optional[str] = None


class WorkspaceMemberUpdate(BaseModel):
    """Update member fields"""
    role: Optional[MemberRole] = None
    permissions: Optional[dict[str, bool]] = None
    status: Optional[MemberStatus] = None


# Invitation Models

class InvitationCreate(BaseModel):
    """Create invitation"""
    email: EmailStr
    role: MemberRole = MemberRole.VIEWER
    message: Optional[str] = None
    expires_in_days: int = 7


class Invitation(BaseModel):
    """Workspace invitation"""
    id: str
    workspace_id: str
    workspace_name: str
    email: str
    role: MemberRole
    status: InvitationStatus
    message: Optional[str] = None
    token: str
    invited_by_id: str
    invited_by_name: str
    expires_at: datetime
    accepted_at: Optional[datetime] = None
    declined_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class BulkInvitationCreate(BaseModel):
    """Bulk invitation creation"""
    emails: list[EmailStr]
    role: MemberRole = MemberRole.VIEWER
    message: Optional[str] = None
    expires_in_days: int = 7


class InvitationResponse(BaseModel):
    """Response for invitation action"""
    accept: bool
    user_id: Optional[str] = None  # For existing users


# Workspace Quota Models

class WorkspaceQuota(BaseModel):
    """Workspace resource quotas"""
    workspace_id: str
    plan: WorkspacePlan
    # Member limits
    max_members: int
    current_members: int
    max_guests: int
    current_guests: int
    # Resource limits
    max_connections: int
    current_connections: int
    max_dashboards: int
    current_dashboards: int
    max_charts: int
    current_charts: int
    max_datasets: int
    current_datasets: int
    # Storage limits
    max_storage_mb: int
    current_storage_mb: float
    # API limits
    max_api_calls_per_day: int
    current_api_calls_today: int
    max_query_execution_minutes: int
    current_query_minutes_today: float
    # Feature limits
    max_scheduled_reports: int
    current_scheduled_reports: int
    max_alerts: int
    current_alerts: int


class QuotaUsage(BaseModel):
    """Real-time quota usage"""
    workspace_id: str
    resource_type: str
    limit: int
    used: int
    remaining: int
    percent_used: float


# Access Control Models

class ResourcePermission(BaseModel):
    """Resource-level permission"""
    resource_type: ResourceType
    resource_id: str
    member_id: str
    can_view: bool = True
    can_edit: bool = False
    can_delete: bool = False
    can_share: bool = False
    can_export: bool = True


class RolePermissions(BaseModel):
    """Permissions for a role"""
    role: MemberRole
    # Workspace permissions
    can_manage_workspace: bool = False
    can_manage_members: bool = False
    can_manage_billing: bool = False
    can_manage_settings: bool = False
    # Resource permissions
    can_create_connections: bool = False
    can_create_dashboards: bool = True
    can_create_charts: bool = True
    can_create_datasets: bool = False
    can_execute_queries: bool = True
    can_export_data: bool = True
    can_share_resources: bool = True
    can_create_public_links: bool = False
    # Feature permissions
    can_use_ai_features: bool = True
    can_create_alerts: bool = False
    can_schedule_reports: bool = False
    can_use_api: bool = False


# Team Models

class Team(BaseModel):
    """Team within workspace"""
    id: str
    workspace_id: str
    name: str
    description: Optional[str] = None
    color: Optional[str] = None
    member_ids: list[str] = Field(default_factory=list)
    permissions: dict[str, bool] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class TeamCreate(BaseModel):
    """Create team"""
    name: str
    description: Optional[str] = None
    color: Optional[str] = None
    member_ids: list[str] = Field(default_factory=list)


class TeamUpdate(BaseModel):
    """Update team"""
    name: Optional[str] = None
    description: Optional[str] = None
    color: Optional[str] = None
    member_ids: Optional[list[str]] = None
    permissions: Optional[dict[str, bool]] = None


# Activity Models

class WorkspaceActivity(BaseModel):
    """Workspace activity log"""
    id: str
    workspace_id: str
    user_id: str
    user_name: str
    action: str
    resource_type: Optional[ResourceType] = None
    resource_id: Optional[str] = None
    resource_name: Optional[str] = None
    details: dict[str, Any] = Field(default_factory=dict)
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


# Response Models

class WorkspaceListResponse(BaseModel):
    """List workspaces response"""
    workspaces: list[Workspace]
    total: int


class WorkspaceMemberListResponse(BaseModel):
    """List members response"""
    members: list[WorkspaceMember]
    total: int


class InvitationListResponse(BaseModel):
    """List invitations response"""
    invitations: list[Invitation]
    total: int


class TeamListResponse(BaseModel):
    """List teams response"""
    teams: list[Team]
    total: int


class WorkspaceActivityListResponse(BaseModel):
    """List activity response"""
    activities: list[WorkspaceActivity]
    total: int


class QuotaUsageListResponse(BaseModel):
    """List quota usage response"""
    quotas: list[QuotaUsage]


# Summary Models

class WorkspaceSummary(BaseModel):
    """Workspace summary for dashboard"""
    workspace_id: str
    name: str
    plan: WorkspacePlan
    status: WorkspaceStatus
    member_count: int
    dashboard_count: int
    connection_count: int
    chart_count: int
    storage_used_mb: float
    storage_limit_mb: int
    api_calls_today: int
    active_users_today: int
    recent_activities: list[WorkspaceActivity]


class UserWorkspaces(BaseModel):
    """User's workspaces"""
    owned: list[Workspace]
    member_of: list[Workspace]
    pending_invitations: list[Invitation]


# Constants

PLAN_LIMITS = {
    WorkspacePlan.FREE: {
        "max_members": 3,
        "max_guests": 0,
        "max_connections": 2,
        "max_dashboards": 5,
        "max_charts": 20,
        "max_datasets": 5,
        "max_storage_mb": 100,
        "max_api_calls_per_day": 100,
        "max_query_execution_minutes": 10,
        "max_scheduled_reports": 0,
        "max_alerts": 0,
    },
    WorkspacePlan.STARTER: {
        "max_members": 5,
        "max_guests": 5,
        "max_connections": 5,
        "max_dashboards": 20,
        "max_charts": 100,
        "max_datasets": 20,
        "max_storage_mb": 500,
        "max_api_calls_per_day": 1000,
        "max_query_execution_minutes": 60,
        "max_scheduled_reports": 5,
        "max_alerts": 10,
    },
    WorkspacePlan.PRO: {
        "max_members": 25,
        "max_guests": 25,
        "max_connections": 20,
        "max_dashboards": 100,
        "max_charts": 500,
        "max_datasets": 100,
        "max_storage_mb": 5000,
        "max_api_calls_per_day": 10000,
        "max_query_execution_minutes": 300,
        "max_scheduled_reports": 25,
        "max_alerts": 50,
    },
    WorkspacePlan.BUSINESS: {
        "max_members": 100,
        "max_guests": 100,
        "max_connections": 50,
        "max_dashboards": 500,
        "max_charts": 2500,
        "max_datasets": 500,
        "max_storage_mb": 25000,
        "max_api_calls_per_day": 100000,
        "max_query_execution_minutes": 1000,
        "max_scheduled_reports": 100,
        "max_alerts": 200,
    },
    WorkspacePlan.ENTERPRISE: {
        "max_members": -1,  # Unlimited
        "max_guests": -1,
        "max_connections": -1,
        "max_dashboards": -1,
        "max_charts": -1,
        "max_datasets": -1,
        "max_storage_mb": -1,
        "max_api_calls_per_day": -1,
        "max_query_execution_minutes": -1,
        "max_scheduled_reports": -1,
        "max_alerts": -1,
    },
}

ROLE_PERMISSIONS = {
    MemberRole.OWNER: RolePermissions(
        role=MemberRole.OWNER,
        can_manage_workspace=True,
        can_manage_members=True,
        can_manage_billing=True,
        can_manage_settings=True,
        can_create_connections=True,
        can_create_dashboards=True,
        can_create_charts=True,
        can_create_datasets=True,
        can_execute_queries=True,
        can_export_data=True,
        can_share_resources=True,
        can_create_public_links=True,
        can_use_ai_features=True,
        can_create_alerts=True,
        can_schedule_reports=True,
        can_use_api=True,
    ),
    MemberRole.ADMIN: RolePermissions(
        role=MemberRole.ADMIN,
        can_manage_workspace=False,
        can_manage_members=True,
        can_manage_billing=False,
        can_manage_settings=True,
        can_create_connections=True,
        can_create_dashboards=True,
        can_create_charts=True,
        can_create_datasets=True,
        can_execute_queries=True,
        can_export_data=True,
        can_share_resources=True,
        can_create_public_links=True,
        can_use_ai_features=True,
        can_create_alerts=True,
        can_schedule_reports=True,
        can_use_api=True,
    ),
    MemberRole.EDITOR: RolePermissions(
        role=MemberRole.EDITOR,
        can_manage_workspace=False,
        can_manage_members=False,
        can_manage_billing=False,
        can_manage_settings=False,
        can_create_connections=False,
        can_create_dashboards=True,
        can_create_charts=True,
        can_create_datasets=True,
        can_execute_queries=True,
        can_export_data=True,
        can_share_resources=True,
        can_create_public_links=False,
        can_use_ai_features=True,
        can_create_alerts=True,
        can_schedule_reports=True,
        can_use_api=False,
    ),
    MemberRole.VIEWER: RolePermissions(
        role=MemberRole.VIEWER,
        can_manage_workspace=False,
        can_manage_members=False,
        can_manage_billing=False,
        can_manage_settings=False,
        can_create_connections=False,
        can_create_dashboards=False,
        can_create_charts=False,
        can_create_datasets=False,
        can_execute_queries=True,
        can_export_data=True,
        can_share_resources=False,
        can_create_public_links=False,
        can_use_ai_features=True,
        can_create_alerts=False,
        can_schedule_reports=False,
        can_use_api=False,
    ),
    MemberRole.GUEST: RolePermissions(
        role=MemberRole.GUEST,
        can_manage_workspace=False,
        can_manage_members=False,
        can_manage_billing=False,
        can_manage_settings=False,
        can_create_connections=False,
        can_create_dashboards=False,
        can_create_charts=False,
        can_create_datasets=False,
        can_execute_queries=False,
        can_export_data=False,
        can_share_resources=False,
        can_create_public_links=False,
        can_use_ai_features=False,
        can_create_alerts=False,
        can_schedule_reports=False,
        can_use_api=False,
    ),
}


# Helper Functions

def get_plan_limits(plan: WorkspacePlan) -> dict[str, int]:
    """Get limits for a plan."""
    return PLAN_LIMITS.get(plan, PLAN_LIMITS[WorkspacePlan.FREE])


def get_role_permissions(role: MemberRole) -> RolePermissions:
    """Get permissions for a role."""
    return ROLE_PERMISSIONS.get(role, ROLE_PERMISSIONS[MemberRole.VIEWER])


def check_quota(quota: WorkspaceQuota, resource_type: str) -> bool:
    """Check if quota allows adding resource."""
    limits = {
        "members": (quota.max_members, quota.current_members),
        "guests": (quota.max_guests, quota.current_guests),
        "connections": (quota.max_connections, quota.current_connections),
        "dashboards": (quota.max_dashboards, quota.current_dashboards),
        "charts": (quota.max_charts, quota.current_charts),
        "datasets": (quota.max_datasets, quota.current_datasets),
        "storage": (quota.max_storage_mb, quota.current_storage_mb),
    }

    if resource_type not in limits:
        return True

    max_val, current = limits[resource_type]
    if max_val == -1:  # Unlimited
        return True

    return current < max_val


def calculate_quota_usage(max_val: int, current: float) -> QuotaUsage:
    """Calculate quota usage percentage."""
    if max_val == -1:
        return QuotaUsage(
            workspace_id="",
            resource_type="",
            limit=-1,
            used=int(current),
            remaining=-1,
            percent_used=0,
        )

    remaining = max_val - int(current)
    percent = (current / max_val * 100) if max_val > 0 else 0

    return QuotaUsage(
        workspace_id="",
        resource_type="",
        limit=max_val,
        used=int(current),
        remaining=remaining,
        percent_used=round(percent, 2),
    )
